src/netconf_acm.c

```diff
@@ -0,0 +1,1833 @@
+                    }
+                }
+            } else if (!strcmp(node->schema->name, "action")) {
+                if (!strcmp(lyd_get_value(node), "permit")) {
+                    rule->action_deny = 0;
+                } else {
+                    rule->action_deny = 1;
+                }
+            } else if (!strcmp(node->schema->name, "comment")) {
+                if (op == SR_OP_DELETED) {
+                    free(rule->comment);
+                    rule->comment = NULL;
+                } else {
+                    assert((op == SR_OP_MODIFIED) || (op == SR_OP_CREATED));
+                    free(rule->comment);
+                    rule->comment = strdup(lyd_get_value(node));
+                }
+            }
+        }
+    }
+
+    /* NACM UNLOCK */
+    pthread_mutex_unlock(&nacm.lock);
+
+    sr_free_change_iter(iter);
+    if (rc != SR_ERR_NOT_FOUND) {
+        sr_session_set_error_message(session, "Getting next change failed (%s).", sr_strerror(rc));
+        return rc;
+    }
+
+    return SR_ERR_OK;
+}
+
+API void
+sr_nacm_init(void)
+{
+    pthread_mutex_init(&nacm.lock, NULL);
+}
+
+#define SR_CONFIG_SUBSCR(session, mod_name, xpath, cb) \
+    rc = sr_module_change_subscribe(session, mod_name, xpath, cb, NULL, 0, \
+            SR_SUBSCR_DONE_ONLY | SR_SUBSCR_ENABLED, sub); \
+    if (rc != SR_ERR_OK) { \
+        sr_errinfo_new(&error_info, rc, "Subscribing for \"%s\" data changes failed.", mod_name); \
+        return error_info; \
+    }
+
+#define SR_OPER_SUBSCR(session, mod_name, xpath, cb) \
+    rc = sr_oper_get_subscribe(session, mod_name, xpath, cb, NULL, 0, sub); \
+    if (rc != SR_ERR_OK) { \
+        sr_errinfo_new(&error_info, rc, "Subscribing for providing \"%s\" state data failed.", mod_name); \
+        return error_info; \
+    }
+
+API sr_error_info_t *
+sr_nacm_init_callbacks(sr_session_ctx_t *session, sr_subscription_ctx_t **sub)
+{
+    sr_error_info_t *error_info = NULL;
+    const char* mod_name;
+    const char* xpath;
+    int rc;
+
+    mod_name = "ietf-netconf-acm";
+    xpath = "/ietf-netconf-acm:nacm";
+    SR_CONFIG_SUBSCR(session, mod_name, xpath, sr_nacm_nacm_params_cb);
+
+    xpath = "/ietf-netconf-acm:nacm/groups/group";
+    SR_CONFIG_SUBSCR(session, mod_name, xpath, sr_nacm_group_cb);
+
+    xpath = "/ietf-netconf-acm:nacm/rule-list";
+    SR_CONFIG_SUBSCR(session, mod_name, xpath, sr_nacm_rule_list_cb);
+
+    xpath = "/ietf-netconf-acm:nacm/rule-list/rule";
+    SR_CONFIG_SUBSCR(session, mod_name, xpath, sr_nacm_rule_cb);
+
+    /* state data */
+    xpath = "/ietf-netconf-acm:nacm/denied-operations";
+    SR_OPER_SUBSCR(session, mod_name, xpath, sr_nacm_oper_cb);
+
+    xpath = "/ietf-netconf-acm:nacm/denied-data-writes";
+    SR_OPER_SUBSCR(session, mod_name, xpath, sr_nacm_oper_cb);
+
+    xpath = "/ietf-netconf-acm:nacm/denied-notifications";
+    SR_OPER_SUBSCR(session, mod_name, xpath, sr_nacm_oper_cb);
+
+    return error_info;
+}
+
+API void
+sr_nacm_destroy(void)
```

Comment:
Just to be safe, it should be possible to call this function several times, so please zero all the freed data.

Reply:
done